pkg/volume/csi/csi_mounter.go

```diff
@@ -206,31 +206,31 @@ func (c *csiMountMgr) SetUpAt(dir string, fsGroup *int64) error {
 	}
 
 	// apply volume ownership
-	if !c.readOnly && fsGroup != nil {
-		err := volume.SetVolumeOwnership(c, fsGroup)
-		if err != nil {
-			// attempt to rollback mount.
-			glog.Error(log("mounter.SetupAt failed to set fsgroup volume ownership for [%s]: %v", c.volumeID, err))
-			glog.V(4).Info(log("mounter.SetupAt attempting to unpublish volume %s due to previous error", c.volumeID))
-			if unpubErr := csi.NodeUnpublishVolume(ctx, c.volumeID, dir); unpubErr != nil {
-				glog.Error(log(
-					"mounter.SetupAt failed to unpublish volume [%s]: %v (caused by previous NodePublish error: %v)",
-					c.volumeID, unpubErr, err,
-				))
-				return fmt.Errorf("%v (caused by %v)", unpubErr, err)
-			}
+	// The following logic is derived from https://github.com/kubernetes/kubernetes/issues/66323
+	// if fstype is "", then skip fsgroup (could be indication of non-block filesystem)
+	// if fstype is provided and pv.AccessMode == ReadWriteOnly, then apply fsgroup
 
-			if unmountErr := removeMountDir(c.plugin, dir); unmountErr != nil {
-				glog.Error(log(
-					"mounter.SetupAt failed to clean mount dir [%s]: %v (caused by previous NodePublish error: %v)",
-					dir, unmountErr, err,
-				))
-				return fmt.Errorf("%v (caused by %v)", unmountErr, err)
-			}
+	err = applyFSGroup(c, fsType, fsGroup)
+	if err != nil {
+		// attempt to rollback mount.
+		glog.Error(log("mounter.SetupAt failed to set fsgroup volume ownership for [%s]: %v", c.volumeID, err))
```

Comment:
👍

Reply:
correct, we only use access modes on the pv in the volume plugin.  For dynamic provisioning, pv access mode will be set to pvc access mode, however for precreated PVs, access modes could contain all the possible modes that the backend supports.

Reply:
@msau42 
Maybe there are couple issues to consider here:

1) Can `Plugin.Mounter` get access to the PV?  Whether generated from PVC or by hand, the PV should be available during the `Mounter.SetupAt` call - right?

2) What do if user provides a hand-created PV that has all possible modes.

I am not sure how to get around number 2 for CSI when that happens.  There is no way to query the driver to get supported Attributes for the volume.  

Wait a minute, we have the `CSIDriver` object now. Maybe a follow up PR is to update the CSIDriver to include supported volumeAttributes for the driver.

PTAL at the updates.

Reply:
1. not an issue.  The code today is using PV.accessmodes
2. I don't see a great way to handle this.  The proper way needs to get access modes from the PVC.  Getting attributes from the driver doesn't help because the driver doesn't know how the user wants to use it.

Reply:
What happens if we drop accessMode check for now and revisit this problem in a follow up github issue?

Reply:
@gnufied good question? Is the way it's implemented now in this PR (with accessmode) a breaking change, does it make it better, or worse? If this PR is not making it better, sure we should kick it down the road. But that would be a fix beyond CSI, right.